02-peer-protocol.md

```diff
@@ -2337,6 +2351,19 @@ A node:
     do not match the expected values):
       - SHOULD send an `error` and fail the channel.
 
+A receiving node:
+  - if `next_funding_txid` is set:
+    - if `next_funding_txid` matches the latest interactive funding transaction:
+      - if it has not received `tx_signatures` for that funding transaction:
+        - MUST retransmit its `commitment_signed` for that funding transaction.
+      - if it should sign first, as specified in the [`tx_signatures` requirements](#the-tx_signatures-message):
+        - MUST send its `tx_signatures` for that funding transaction.
+      - if it has already received `tx_signatures` for that funding transaction:
+        - MUST send its `tx_signatures` for that funding transaction.
```

Comment:
The other node may not have actually sent their `commitment_signed` previously, or it may have been lost.  In that case, we MUST NOT send our `tx_signatures` until they have retransmitted `commitment_signed`.

```suggestion
      - if it has not received both `commitment_signed` and `tx_signatures` for that funding transaction:
        - MUST retransmit its `commitment_signed` for that funding transaction.
      - if it should sign first, as specified in the [`tx_signatures` requirements](#the-tx_signatures-message):
        - if it has already received `commitment_signed`:
          - MUST send its `tx_signatures` for that funding transaction.
        - otherwise:
          - MUST send its `tx_signatures` for that funding transaction ONCE a valid `commitment_signed` has been received.
          - MUST NOT send its `tx_signatures` UNTIL a valid `commitment_signed` has been received.
      - if it has already received both `commitment_signed` and `tx_signatures` for that funding transaction:
        - MUST send its `tx_signatures` for that funding transaction.
```

Reply:
Good catch, the previous version didn't match the actual code I wrote (writing specification is harder than writing code!). But I believe we can keep the requirements simpler, what I did in https://github.com/niftynei/lightning-rfc/pull/13/commits/b6c4c7c9666bf7093b65bf451b4b38e81a750b4d should now exactly match what I implemented and should easily map to nested ifs.

I don't think we need to explicitly say "received both `commitment_signed` and `tx_signatures`": if you've received `tx_signatures`, it means you've also received `commitment_signed` and your peer has received your `commitment_signed`. The condition in the code itself only needs to look at whether we received `tx_signatures` or not. But you're totally right that before sending our `tx_signatures`, we need to make sure we've received the remote `commitment_signed`!